tests/kernel/smp/src/main.c

```diff
@@ -83,9 +84,48 @@ void test_main(void)
 		}
 	}
 
-	if (ok) {
-		TC_END_REPORT(TC_PASS);
-	} else {
-		TC_END_REPORT(TC_FAIL);
-	}
+	zassert_true(ok, "SMP test failed");
+}
+
+static void child_fn(void *p1, void *p2, void *p3)
+{
+	ARG_UNUSED(p3);
+	int parent_cpu_id = (int)p1;
+	struct k_thread *parent_thread_id = (struct k_thread *)p2;
+
+	zassert_true(parent_cpu_id != _arch_curr_cpu()->id &&
+		     _kernel.cpus[parent_cpu_id].current
+		     == parent_thread_id,
+		     "Parent isn't on other core");
```

Comment:
This is going to be as reliable as you hope.  You don't know for sure that there aren't other threads (work queues, maybe) that might be preempting either CPU.  There probably aren't right now, in this test, but tests get augmented and moved around all the time and in some of the bigger ones there are MANY "forgotten" threads floating around from "finished" tests that still have a little work to do (and like to exercise scheduler bugs, heh!).

Broadly: Whiteboxing _kernel internals isn't exactly unacceptable in tests, but I don't see why this is needed here.  If you want to see that the other CPU is running the thread you expect, have it do something.  For example you could have one spin on an atomic flag variable (just a volatile global int is fine) waiting for a "message" from the other CPU, then set some output flag in a particular way.  Then you can know what is happening on the other CPU without mucking with internals.